Making AssetManager more "batteries-included" by default

mechaxl opened the pull request.

mechaxl commented on examples/06_assets/main.rs:
```diff
@@ -1,66 +1,32 @@
-use cgmath::{InnerSpace, Vector3};
-use std::io::BufReader;
-
-struct Obj(obj::Obj);
```
I'm wondering if there's any rationale behind keeping this here. Loading `*.obj` files seems like a a pretty common use case, so it made sense to me to move this directly to the asset manager module, but it could certainly be moved back to here.

Emilgardis replied:
I feel like there should still be an example showing how to implement a custom asset loader. Or maybe it's too trivial

anderejd replied:
I think the goal of the pull request is excellent, more standard formats supported by the asset manger would be fantastic. I also agree with @Emilgardis that there shoud be an example demonstrating custom asset loading. Maybe do both in 06_assets? Move obj loader code to asset mananger and implement some obviously custom format loader in the example code and load one obj asset and one custom asset?

diegor8 replied:
could give your thoughts about this please

mechaxl commented on examples/06_assets/main.rs:
```diff
@@ -117,6 +94,10 @@ impl State for Example {
 fn main() {
+    let assets_path = format!("{}/examples/06_assets/resources/assets",
+                   env!("CARGO_MANIFEST_DIR"));
+    set_var("AMETHYST_ASSET_DIRS", assets_path);
```
Also added some handling for setting up default asset directories. Do we want to demo that here, or revert this to setting up a `DirectoryStore` itself? It does seem kind of odd to have this example setting up asset loading differently from the others.

Emilgardis replied:
I think this should be done by the user, also the "static" set env isn't that great. But I feel like this behaviour should be kept somehow optionally.

diegor8 commented:
my issue may be of interest to the people of this thread

https://github.com/amethyst/amethyst/issues/150

mechaxl commented:
@Emilgardis, @anderejd, @msiglreith: I've rebased this PR off of the latest changes to `develop`, and added a custom object loader example. It should be ready for another review when you get the chance. Thanks!

Aceeri commented on Cargo.toml:
```diff
@@ -49,13 +49,14 @@ name = "assets"
-genmesh="0.4"
+glutin = "0.6"
+imagefmt = "*"
+wavefront_obj = "*"
```
These shouldn't be wildcarded, could break the example in the future if the dependent crates publish a major change.

mechaxl replied:
:+1: Updated

Aceeri commented on examples/06_assets/main.rs:
```diff
@@ -59,62 +87,105 @@ impl State for Example {
+        let renderable = asset_manager.create_renderable("sphere", "grass", "green").unwrap();
+        let mut transform = LocalTransform::default();
+        transform.translation = [-5.0, 0.0, 7.5];
+        transform.rotation = [0.70711, 0.70711, 0.0, 0.0];
```
Is this correct? The rotation is a quaternion [w, x, y, z] so I'm thinking this might need to be `[1.0, 0.70711, 0.70711, 0.0]` if you wanted a 45 degree angle on x and y.

Emilgardis replied:
Is this resolved?

mechaxl replied:
I think so. @aceeri?

Aceeri replied:
I suppose, just seems weird to me.

mechaxl replied:
@Aceeri: With that rotation, I'm getting this result, which I'm guessing is not correct?

<img width="1022" alt="rotation" src="https://cloud.githubusercontent.com/assets/765788/21599227/fcebd0a4-d12e-11e6-9573-2e6b25b82af9.png">

I was mostly just trying to point the top of the sphere straight up and away from the camera, so I ended up rotating it 90 degrees around the X axis.

msiglreith replied:
Would prefer using cgmath functions to build the quaternions from axis rotations or euler angles, as they are more intuitive when reading the code.
Note: The quaternion is valid (45° around x-axis), but shouldn't make any difference for a sphere compared to not rotating it at all.

mechaxl commented:
@kvark: Looks like you're involved in the `gfx` project. Any idea what's happening with this?

https://github.com/gfx-rs/gfx/issues/1115

It's affecting the ability to run the example with just `cargo run`. `cargo run --release` works just fine, however.

kvark commented:
@mechaxl we'll investigate, hold on

msiglreith commented on src/asset_manager/asset_manager.rs:
```diff
@@ -280,6 +320,101 @@ impl AssetStore for DirectoryStore {
+impl AssetLoader<Texture> for imagefmt::Image<u8> {
+    fn from_data(assets: &mut Assets, image: imagefmt::Image<u8>) -> Option<Texture> {
+        let pixels = image.buf.chunks(4).map(|p| [p[0], p[1], p[2], p[3]]).collect::<Vec<_>>();
+        let chunked = pixels.as_slice().chunks(image.w).collect::<Vec<_>>();
```
`raw` in `TextureLoadData` are slices on images/textures. This is not a 2d array of image data!
For data loaded from an image this means a single slice of pixel data.

This should also fix the mipmap issue:
```Rust
let pixels = image.buf.chunks(4).map(|p| [p[0], p[1], p[2], p[3]]).collect::<Vec<_>>();
AssetLoader::from_data(assets, TextureLoadData {
            kind: Kind::D2(image.w as u16, image.h as u16, AaMode::Single),
            raw: &[pixels.as_slice()],
       })

mechaxl replied:
👍 Thanks! I've fixed the texture loading, and added some comments to the `TextureLoadData` struct to help prevent anyone else from making the same mistake

mechaxl commented:
@msiglreith: I've updated the rotations to use `cgmath`

kvark commented:
@mechaxl for the future, please consider writing a good PR description, setting the context for future reviews/discussions.

kvark commented:
I looked at the code, seems fine to me. @Aceeri approved, so if @msiglreith is also happy, we'll merge
